pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/DirectProxyHandler.java

```diff
@@ -0,0 +1,181 @@
+        b.handler(new ChannelInitializer<SocketChannel>() {
+            @Override
+            protected void initChannel(SocketChannel ch) throws Exception {
+                ch.pipeline().addLast("frameDecoder",
+                        new PulsarLengthFieldFrameDecoder(PulsarDecoder.MaxFrameSize, 0, 4, 0, 4));
+                ch.pipeline().addLast(new ProxyBackendHandler());
+            }
+        });
+
+        URI targetBroker;
+        try {
+            // targetBrokerUrl is coming in the "hostname:6650" form, so we need to extract host and port
+            targetBroker = new URI("pulsar://" + targetBrokerUrl);
+        } catch (URISyntaxException e) {
+            log.warn("[{}] Failed to parse broker url '{}'", inboundChannel, targetBrokerUrl, e);
+            inboundChannel.close();
+            return;
+        }
+
+        ChannelFuture f = b.connect(targetBroker.getHost(), targetBroker.getPort());
+        outboundChannel = f.channel();
+        f.addListener(future -> {
+            if (!future.isSuccess()) {
+                // Close the connection if the connection attempt has failed.
+                inboundChannel.close();
+            }
+        });
+    }
+
+    enum BackendState {
+        Init, HandshakeCompleted
+    }
+
+    public class ProxyBackendHandler extends PulsarDecoder implements FutureListener<Void> {
+
+        private BackendState state = BackendState.Init;
+
+        @Override
+        public void channelActive(ChannelHandlerContext ctx) throws Exception {
+            // Send the Connect command to broker
+            String authData = "";
+            if (authentication.getAuthData().hasDataFromCommand()) {
+                authData = authentication.getAuthData().getCommandData();
+            }
+            outboundChannel
+                    .writeAndFlush(Commands.newConnect(authentication.getAuthMethodName(), authData, "Pulsar proxy"));
+            outboundChannel.read();
+        }
+
+        @Override
+        public void channelRead(final ChannelHandlerContext ctx, Object msg) throws Exception {
+            switch (state) {
+            case Init:
+                if (log.isDebugEnabled()) {
+                    log.debug("[{}] [{}] Received msg on broker connection: {}", inboundChannel, outboundChannel,
+                            msg.getClass());
+                }
+
+                // Do the regular decoding for the Connected message
+                super.channelRead(ctx, msg);
+                break;
+
+            case HandshakeCompleted:
+                inboundChannel.writeAndFlush(msg).addListener(this);
+                break;
+
+            default:
+                break;
+            }
+
+        }
+
+        @Override
+        public void operationComplete(Future<Void> future) throws Exception {
+            // This is invoked when the write operation on the paired connection is completed
+            if (future.isSuccess()) {
+                outboundChannel.read();
```

Comment:
No, the auto-read on the channel is set to false. The reason is to adapt to the speed of the slowest connection. 

So, we read 1 buffer from `c1`, write into `c2`. When the write is completed (meaning we were able to write on the socket because the TCP send window was open), then we read the next buffer.